### DISCUSSION TITLE
DB Scheme possible with Pocketbase?

### DISCUSSION THREAD

**Discussion by FNG-2002** [pocketbase/pocketbase#1997] [Q&A]
Hello everyone, I wanted to share my current db schema and would like to discuss with you how I can implement this into Pocketbase.

I have the following db model (some fields are in german. if that's confusing I can change it later):
![grafik](https://user-images.githubusercontent.com/36012999/223455496-02c2b8e0-070e-4f9a-a7e0-b4e26e08ce30.png)

There are birds and birdfamilies. A user can capture a bird. A bird can be captured by multiple users.
A user, that is public, can share their captured progress and others should be allowed to see it.

I implemented it like this into Pocketbase. Now I want to query the captured birds for a specific user grouped by the username and family name. I was thinking about a response like this:
```
{
  "username": "john",
  "capturedPerFamily": [
    {
      "family": "A",
      "captured": 2,
      "capturedBirds": [
        {
          "name": "Bird1",
          ...
        },
        ...
      ]
    },
    ...
  ]
}

```

On a custom backend I would gather the information on function call getCapturedBirdsByUser("john"). Inside I probably would query the needed information and define a response type and return it.

How would I do this with pocketbase? Is it possible to achieve something like this?

------- My attempt ------- 

I created a view in pocketbase with following SQL Statement
```
SELECT (ROW_NUMBER() OVER()) as id,
  t.username, json_group_array(t.test) AS data
  FROM (
    SELECT u.username AS username,
      json_object(
        'familie',
        f.name,
        'data',
        json_group_array(
          json_object('art', a.name, 'gesehen', au.id NOTNULL)
        )
      ) AS test
    FROM users AS u
      JOIN familie f
      INNER JOIN art a ON a.familie = f.id
      LEFT JOIN art_user au ON au.art = a.id
      AND au.user = u.id
    GROUP BY u.username,
      f.name
) AS t
GROUP BY t.username;
```

The response is following

```
{
...
  "items": [
    {
      "collectionId": "...",
      "collectionName": "...",
      "id": "1",
      "data": [
        "{\"familie\":\"Tauben\",\"data\":[{\"art\":\"Buntspecht\",\"gesehen\":1}]}",
        "{\"familie\":\"Test\",\"data\":[{\"art\":\"Schwan\",\"gesehen\":0},{\"art\":\"Test\",\"gesehen\":0}]}"
      ],
      "username": "ada"
    },
    {
      "collectionId": "...",
      "collectionName": "...",
      "id": "2",
      "data": [
        "{\"familie\":\"Tauben\",\"data\":[{\"art\":\"Buntspecht\",\"gesehen\":1}]}",
        "{\"familie\":\"Test\",\"data\":[{\"art\":\"Schwan\",\"gesehen\":1},{\"art\":\"Test\",\"gesehen\":0}]}"
      ],
      "username": "john"
    }
  ]
}

```

I mean this kinda works but it really feels not good doing it like this. Also I have the problem, that this view contains this data for *every* user. For my purpose I could filter it, but someone calling the url without the filter could potentially get all the information for the other users in one call. Assume I have 1000 users who each captured lots of birds this would be a lot of data.

I hope I didn't overwhelm any of you. I'm really interested what your point of view is.
Thanks in advance ^^

**Comment by ganigeorgiev** [comment]
>  On a custom backend I would gather the information on function call getCapturedBirdsByUser("john"). Inside I probably would query the needed information and define a response type and return it.
> 
> How would I do this with pocketbase? Is it possible to achieve something like this?


I'm sorry, I don't understand what you want to do. If you haven't already, please read https://pocketbase.io/docs/collections/ to get a general idea how PocketBase data are structured.


The provided view query is also confusing. Please elaborate what you are trying to do and what is the expected result. A simplified schema/example would be also helpful.

> For my purpose I could filter it, but someone calling the url without the filter could potentially get all the information for the other users in one call.

Collection API rules (that includes also views) act as both security rules and filters, aka. you can define List/Search or View API rule with the specified filter. You can find some information in https://pocketbase.io/docs/api-rules-and-filters/.

**Reply by ganigeorgiev** [reply]
Additionally, instead of manually serializing the related records in a view, for expanding relations you can check https://pocketbase.io/docs/expanding-relations/.

**Reply by FNG-2002** [reply]
Thanks for your response. I'm sorry. I tried to simplify the problem.
Here's my simplified schema:
![grafik](https://user-images.githubusercontent.com/36012999/223466445-70ee82d5-c796-4fe5-91d0-cdbc2527b532.png)

I also rethought the view and have following now
![grafik](https://user-images.githubusercontent.com/36012999/223477513-2b343dc7-d5d5-4a85-8699-9cae0d057c65.png)
So i got the birds grouped by the user and familiy. Now if I expand and filter with `?expand=user&filter=(user.username="ada")` I can get the captured birds by ada grouped by the family.

I got two problems/questions. 
1. Now I got the captured birds (their ids) in an array and cant expand them. Is there any way to expand them too?
2. The user has a "public" field. Only public users should be exposed. I know that I can define a rule like `user.public = true`. But if I call the API Endpoint without a filter for a user I get the captured birds grouped by familiy for all (public) users. Is this something desireable? I actually want to force filter this view, so it cannot be used without a filter. Any ideas how to do this?

Hopefully it's clearer now :) If not please let me know.

**Reply by dschissler** [reply]
> Now I got the captured birds (their ids) in an array and cant expand them. Is there any way to expand them too?

Make sure that the user has access to the expanded collections in the API Rules for View or List.

**Reply by ganigeorgiev** [reply]
@FNG-2002 

> 1. Now I got the captured birds (their ids) in an array and cant expand them. Is there any way to expand them too?

From the screenshot the `arten` column seems to be a SQL expression, which will be detected as `json` type field, but `expand` works only with `relation` type fields, so for this particular case you can't use `expand`. You'll have to send separate request(s) to fetch the extracted `arten` records (eg. using a filter with `id="id1" || id="id2" || etc.`) and then eventually group them client-side.

> 2. Is this something desireable? I actually want to force filter this view, so it cannot be used without a filter. Any ideas how to do this?

It is a little hacky, but one workaround to enforce a dynamic `filter` could be to define an API rule like: `@request.query.filter ~ 'user.username'`.

---

I think I understand a little better what query you want, but I'm not sure what to recommend since it really depends how you plan to use the grouped data. Maybe something similar to your initial approach with `json_group_array` and `json_object` could be better for this particular use case. I'm not really sure.
Alternatively, if you need more advanced control and crafted data responses, you could [register a custom route](https://pocketbase.io/docs/router/#register-new-route) and [manually query the database](https://pocketbase.io/docs/database/).

**Reply by dschissler** [reply]
@ganigeorgiev I don't exactly follow, but would you also recommend a new View collection for this?  Is there a limitation there around unpacking json?

**Reply by ganigeorgiev** [reply]
@dschissler

> I don't exactly follow, but would you also recommend a new View collection for this? Is there a limitation there around unpacking json?

The limitation at the moment is that in the above the SQL column expression is detected as `json` field instead of multiple `relation` field. The reason for that is because it is difficult to properly detect what a SQL expression do without implementing a full-blown SQL parser, which I want to avoid. And often actually it may not be even possible because the array values could be technically anything and there is no guarantee that all values are ids from the same collection (eg. the array could be a result of concatenated value from multiple different columns).

One workaround for this could be to allow users to enforce a view schema in addition to the autogenerated one, which will enable for example in the above to manually "adjust" the `arten` column to a multiple `relation` field with a specified collection, but it seems error prone and I'll have to think a little more on it.

**Reply by FNG-2002** [reply]
> It is a little hacky, but one workaround to enforce a dynamic `filter` could be to define an API rule like: `@request.query.filter ~ 'user.username'`.

That's really cool, I didn't think about that. Thank you!

> I think I understand a little better what query you want, but I'm not sure what to recommend since it really depends how you plan to use the grouped data. 

Yes, that's true. It really depends. For now I don't have a frontend, I'm just playing around with pocketbase.

>Alternatively, if you need more advanced control and crafted data responses, you could [register a custom route](https://pocketbase.io/docs/router/#register-new-route) and [manually query the database](https://pocketbase.io/docs/database/).

Thanks, I will try this too.

**Reply by FNG-2002** [reply]
Update for anyone that stumbles over this

> It is a little hacky, but one workaround to enforce a dynamic `filter` could be to define an API rule like: `@request.query.filter ~ 'user.username'`.

I got it working with `@request.query.filter ~ 'user.username="%"'`
The problem I got with `@request.query.filter ~ 'user.username'` is that `?filter=(user.username!='x')` avoids it and would give me the data for all users.
